doc/design/programming_solver.md

```diff
@@ -0,0 +1,280 @@
+
+To solve this problem, we can use tools like [AMPL](https://en.wikipedia.org/wiki/AMPL) or open-source tools like [pyomo](http://www.pyomo.org/), or [Matlab](https://www.mathworks.com/help/optim/ug/linprog.html), or [R](https://towardsdatascience.com/linear-programming-in-r-444e9c199280).
+
+You can learn from the examples in the above links that:
+
+- Using Matlab and R are quite the same, they require users to define the constraints as matrixes and call a function to get the result. They both have their own grammar, and you have to write code according to their language specifications.
+- Using "AMPL like" high-level language to describe the problem, and call the corresponding solvers. Pyomo and AMPL have similar components in their grammar: sets, parameters, variables, objectives, constraints (https://pyomo.readthedocs.io/en/stable/pyomo_modeling_components/index.html).
+
+So we can see that using AMPL is a modern and general way of describing mathematical programming problems. We can simply write the AMPL snippet to describe the above problem:
+
+```
+set X;
+var ProductAmt{x in X} >= 0;
+
+param Price{x in X};
+param MaterialCost{x in X};
+param OtherCost{x in X};
+param Finishing{x in X};
+param Carpentry{x in X};
+param Demand{x in X};
+
+maximize profit:
+    sum{x in X} (Price[x] - MaterialCost[x] - OtherCost[x]) * ProductAmt[x];
+
+s.t. finishing: sum{x in X} Finishing[x] * ProductAmt[x] <= 100;
+s.t. carpentry: sum{x in X} Carpentry[x] * ProductAmt[x] <= 80;
+s.t. demand{x in X}: ProductAmt[x] <= 40;
+```
+
+## Grammar Design
+
+In order to extend SQL to have completely same ability of AMPL, the extended syntax should be able to describe **objective and constraints** while the input data table can store the **params** for each **variable**, and the rows in the table is naturally become the **set** we defined in AMPL. 
+
+### Single Variable
+
+Then we have the below `train_table`:
+
+| product | price | materials_cost | other_cost | finishing | carpentry | max_num |
+| ------- | ----- | -------------- | ---------- | --------- | --------- | ------- |
+| soldier | 27    | 10             | 14         | 2         | 1         | 40      |
+| train   | 21    | 9              | 10         | 1         | 1         | 10000   |
+
+In the `train_table`:
+
+- The set X is row one and row two.
+- We have one variable, and the variable name strings is stored in column `product`. In cases that have multiple variables (like the example described at https://en.wikipedia.org/wiki/AMPL), the table should have multiple string columns to store the variable.
+- Other columns like `price`, `materials_cost` are all params for the corresponding variable.
+
+Then we can use below extended SQL syntax to describe above example:
+
+```sql
+SELECT * FROM train_table
+TO MAXIMIZE SUM((price - materials_cost - other_cost) * product)
+CONSTRAINT SUM(finishing * product) <= 100
+CONSTRAINT SUM(carpentry * product) <= 80
+CONSTRAINT product <= max_num
```

Comment:
I was wondering whether we can use one word `CONSTRAINT`, just like `WITH`.
```sql
SELECT * FROM train_table TO MAXIMIZE SUM((price - materials_cost - other_cost) * product)
CONSTRAINT SUM(finishing * product) <= 100,
           SUM(carpentry * product) <= 80,
           product <= max_num
...
```

Reply:
Will there be a logical expression among the `CONSTRAINT` clause. Like expression:
``` SQL
CONSTRAINT a or b,
CONSTRAINT c
```
equals to `(a || b) && c`

Reply:
There's not. All constraints must be logical `and`.

Reply:
> I was wondering whether we can use one word CONSTRAINT, just like WITH.

Sounds cool.